include/nbl/builtin/hlsl/emulated/emulated_float64_t.hlsl

```diff
@@ -0,0 +1,622 @@
+#ifndef _NBL_BUILTIN_HLSL_EMULATED_FLOAT64_T_INCLUDED_
+#define _NBL_BUILTIN_HLSL_EMULATED_FLOAT64_T_INCLUDED_
+
+#include <nbl/builtin/hlsl/emulated/emulated_float64_t_impl.hlsl>
+
+namespace nbl
+{
+namespace hlsl
+{
+    template<bool FastMath = true, bool FlushDenormToZero = true>
+    struct emulated_float64_t
+    {
+        using storage_t = uint64_t;
+        using this_t = emulated_float64_t<FastMath, FlushDenormToZero>;
+
+        storage_t data;
+
+        // constructors
+        /*static emulated_float64_t create(uint16_t val)
+        {
+            return emulated_float64_t(bit_cast<uint64_t>(float64_t(val)));
+        }*/
+
+        NBL_CONSTEXPR_STATIC_INLINE this_t create(this_t val)
+        {
+            return val;
+        }
+
+        NBL_CONSTEXPR_STATIC_INLINE this_t create(int32_t val)
+        {
+            return bit_cast<this_t>(impl::castToUint64WithFloat64BitPattern(int64_t(val)));
+        }
+
+        NBL_CONSTEXPR_STATIC_INLINE this_t create(int64_t val)
+        {
+            return bit_cast<this_t>(impl::castToUint64WithFloat64BitPattern(val));
+        }
+
+        NBL_CONSTEXPR_STATIC_INLINE this_t create(uint32_t val)
+        {
+            return bit_cast<this_t>(impl::castToUint64WithFloat64BitPattern(uint64_t(val)));
+        }
+
+        NBL_CONSTEXPR_STATIC_INLINE this_t create(uint64_t val)
+        {
+            return bit_cast<this_t>(impl::castToUint64WithFloat64BitPattern(val));
+        }
+
+        NBL_CONSTEXPR_STATIC_INLINE this_t create(float32_t val)
+        {
+            this_t output;
+            output.data = impl::castFloat32ToStorageType<FlushDenormToZero>(val);
+            return output;
+        }
+
+        NBL_CONSTEXPR_STATIC_INLINE this_t create(float64_t val)
+        {
+#ifdef __HLSL_VERSION
+            emulated_float64_t retval;
+            uint32_t lo, hi;
+            asuint(val, lo, hi);
+            retval.data = (uint64_t(hi) << 32) | lo;
+            return retval;
+#else
+            return bit_cast<this_t>(reinterpret_cast<uint64_t&>(val));
+#endif
+        }
+        
+        // TODO: unresolved external symbol imath_half_to_float_table
+        /*static emulated_float64_t create(float16_t val)
+        {
+            return emulated_float64_t(bit_cast<uint64_t>(float64_t(val)));
+        }*/
+
+        // arithmetic operators
+        this_t operator+(const emulated_float64_t rhs) NBL_CONST_MEMBER_FUNC
+        {
```

Comment:
pro tip, have a `add` method and `subtract` method hidden in `emulated_float64_base`

This way you can forward adds with differing signs and subtracts with matching signs to a potentially expensive `subtract` method (128bit due to your 8 ULP errors), while adds with matching signs and subtracts with differing signs go to an `add` method thats 64bit and fast

Reply:
TODO